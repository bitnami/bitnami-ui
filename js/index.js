/*!
 * Bitnami UI - v{VERSION}
 *
 * This is the main javascript file of Bitnami Pattern lib. It contains all
 * the required JS code for interacting with the library. It's written in
 * pure JS.
 *
 * ----------------
 *
 */

// Define global variables
window.BITNAMI_UI_VERSION = 'v{VERSION}';
window.BITNAMI_UI_INIT = false;

// Import components
import Accordion from './components/accordion';
import Carousel from './components/carousel';
import Header from './components/header';

// Export the B selector to window
import b from './base';
window.b = b;

// Initialize all elements
document.addEventListener('DOMContentLoaded', () => {
  Accordion();
  Carousel();
  Header();
});
