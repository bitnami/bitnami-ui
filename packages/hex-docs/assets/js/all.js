// Selection cursors
const saveCursor = (containerEl) => {
  var range = window.getSelection().getRangeAt(0);
  var preSelectionRange = range.cloneRange();
  preSelectionRange.selectNodeContents(containerEl);
  preSelectionRange.setEnd(range.startContainer, range.startOffset);
  var start = preSelectionRange.toString().length;

  return {
      start: start,
      end: start + range.toString().length
  };
};

const restoreCursor = (containerEl, savedSel) => {
  var charIndex = 0, range = document.createRange();
  range.setStart(containerEl, 0);
  range.collapse(true);
  var nodeStack = [containerEl], node, foundStart = false, stop = false;

  while (!stop && (node = nodeStack.pop())) {
      if (node.nodeType == 3) {
          var nextCharIndex = charIndex + node.length;
          if (!foundStart && savedSel.start >= charIndex && savedSel.start <= nextCharIndex) {
              range.setStart(node, savedSel.start - charIndex);
              foundStart = true;
          }
          if (foundStart && savedSel.end >= charIndex && savedSel.end <= nextCharIndex) {
              range.setEnd(node, savedSel.end - charIndex);
              stop = true;
          }
          charIndex = nextCharIndex;
      } else {
          var i = node.childNodes.length;
          while (i--) {
              nodeStack.push(node.childNodes[i]);
          }
      }
  }

  var sel = window.getSelection();
  sel.removeAllRanges();
  sel.addRange(range);
}

// Make code editable
document.addEventListener('DOMContentLoaded', () => {
  document.querySelectorAll('.Component__Content > pre').forEach((n) => {
    n.contentEditable = "true";
    n.addEventListener('input', (e) => {
      const target = e.target;
      const value = target.innerText;
      // Save cursor
      const cursor = saveCursor(target);

      // Remove div on insertParagraph
      if (e.inputType === 'insertParagraph') {
        target.querySelector('div').remove();
        // Increase the cursor position:
        cursor.start += 1;
        cursor.end += 1;
      }

      // highlight
      const result = Prism.highlight(value, Prism.languages.html);

      // Inject
      target.previousElementSibling.innerHTML = value;
      target.firstChild.innerHTML = result;
      // Restore the cursor
      restoreCursor(target, cursor);
    });
  });

  // Get the current release
  if (b('.currentVersion').length > 0) {
    fetch('https://api.github.com/repos/bitnami/hex/releases/latest')
      .then((res) => res.json())
      .then((data) => {
        b('.currentVersion').html.replace(data.tag_name);
      });
  }
});
