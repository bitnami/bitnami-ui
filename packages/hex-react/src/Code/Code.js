import React from 'react';
import PropTypes from 'prop-types';
import Lowlight from 'lowlight/lib/core';
import Indent from 'detect-indent';

/**
 * The usage of Lowlight in this component is inspired by https://github.com/rexxars/react-lowlight/
 */
class Code extends React.Component {
  constructor(props) {
    super(props);

    // Bind
    this._mapChildren = this._mapChildren.bind(this);
  }

  /**
   * Convert HighlightJS nodes from Lowlight into React component definitions.
   *
   * @param {object} child HighlightJS node from Lowlight
   * @param {number} i Index of the element in the array
   * @param {number} depth Current depth. This is a recursive method
   */
  _mapChildren(child, i, depth = 0) {
    if (child.tagName) {
      return React.createElement(
        child.tagName,
        Object.assign({ key: `lo-${depth}-${i}` }, child.properties),
        child.children && child.children.map((c, j) => this._mapChildren(c, j, depth + 1)),
      );
    }

    return child.value;
  }

  /**
   * Takes the string code set as children and remove the extra indentation due to JSX. It also
   * removes the initial and the final line if they are empty.
   */
  _cleanCode() {
    const { children } = this.props;
    const indentation = Indent(children);
    let minIndent;
    const split = children.split('\n');

    // Remove first line if empty
    if (split[0] === '') {
      split.splice(0, 1);
    }

    // Remove last line if empty
    if (/^\s*$/.test(split[split.length - 1])) {
      split.splice(split.length - 1, 1);
    }

    if (split.length === 1) {
      minIndent = 0;
    } else {
      // Get the minimium indentation. That will be indentation in the JSX file
      split.forEach(line => {
        const { length } = line.match(/^\s*/)[0];
        // Ignore empty lines
        if (line !== '' && (minIndent == null || length < minIndent)) {
          minIndent = length;
        }
      });
    }

    let regexp;

    // Create the regexp to the initial indentation based on the type
    if (indentation.type === 'tab') {
      regexp = new RegExp(`^[\t]{${minIndent}}`, 'mg');
    } else {
      regexp = new RegExp(`^[ ]{${minIndent}}`, 'mg');
    }

    // Join again and remove
    return split.join('\n').replace(regexp, '');
  }

  render() {
    const { highlight, language, className, inline, disableIndent, children } = this.props;
    const childrenCode = inline || disableIndent ? children : this._cleanCode();

    // Highlight the code and format it as React component definitions
    let value;

    if (highlight) {
      const ast =
        language === 'auto'
          ? Lowlight.highlightAuto(childrenCode).value
          : Lowlight.highlight(language, childrenCode).value;

      value =
        ast.length === 0 ? childrenCode : ast.map((child, i) => this._mapChildren(child, i, 0));
    } else {
      value = childrenCode;
    }

    // Inner code tag
    const code = React.createElement('code', {}, value);

    // Display the code for inline or wrap into a pre for blocks
    return inline ? code : React.createElement('pre', { className }, code);
  }
}

Code.propTypes = {
  /**
   * Enable syntax highlighting
   */
  highlight: PropTypes.bool,
  /**
   * Set the language of the code. Use `auto` for autodetection
   */
  language: PropTypes.string,
  /**
   * Base CSS class for the component
   */
  className: PropTypes.string,
  /**
   * String with the actual code we want to highlight
   */
  children: PropTypes.string.isRequired,
  /**
   * Display the code inline
   */
  inline: PropTypes.bool,
  /**
   * Disable the fix of the JSX indentation in the `pre` code block
   */
  disableIndent: PropTypes.bool,
};

Code.defaultProps = {
  highlight: true,
  language: 'auto',
  inline: false,
  className: '',
  disableIndent: false,
};

// Register new languages and other language utilities
Code.registerLanguage = Lowlight.registerLanguage;
Code.hasLanguage = lang => !!Lowlight.getLanguage(lang);

export default Code;
