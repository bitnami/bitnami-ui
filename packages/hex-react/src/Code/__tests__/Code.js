import React from 'react';
import { shallow } from 'enzyme';

// Code block
import js from 'highlight.js/lib/languages/javascript';
import ruby from 'highlight.js/lib/languages/ruby';

import Code from '../Code';

// Register new languages
Code.registerLanguage('js', js);
Code.registerLanguage('ruby', ruby);

describe('Code', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<Code language="js">const test = 1;</Code>);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders multiline correctly', () => {
    const wrapper = shallow(
      <Code language="js">
        {`
          const test = 1;
          const f = (t) => t + 1;
        `}
      </Code>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('register languages correctly', () => {
    expect(Code.hasLanguage('ruby')).toBe(true);
  });

  it('indents the code correctly', () => {
    const wrapper = shallow(
      <Code language="js">
        {`
          const test = 1;
          const f = (t) => t + 1;
        `}
      </Code>,
    );
    const expected = `const test = 1;\nconst f = (t) => t + 1;`;
    expect(wrapper.text()).toEqual(expected);
  });

  it('can disable auto-indentation', () => {
    const wrapper = shallow(
      <Code language="js" disableIndent>
        {`
          const test = 1;
          const f = (t) => t + 1;
        `}
      </Code>,
    );
    const expected = `const test = 1;\nconst f = (t) => t + 1;`;
    expect(wrapper.text()).not.toEqual(expected);
  });

  it('adds pre to blocks', () => {
    const wrapper = shallow(<Code language="js">const test = 1;</Code>);
    expect(wrapper.is('pre')).toBe(true);
  });

  it('uses code directly for inline', () => {
    const wrapper = shallow(
      <Code language="js" inline>
        const test = 1;
      </Code>,
    );
    expect(wrapper.is('code')).toBe(true);
  });

  it('adds extra CSS classes', () => {
    const css = 'test';
    const wrapper = shallow(
      <Code language="js" className={css}>
        const test = 1;
      </Code>,
    );
    expect(wrapper.hasClass(css)).toBe(true);
  });
});
