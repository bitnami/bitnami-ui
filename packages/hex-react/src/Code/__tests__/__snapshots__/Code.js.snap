// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Code renders correctly 1`] = `
<pre
  className=""
>
  <code>
    <span
      className={
        Array [
          "hljs-keyword",
        ]
      }
      key="lo-0-0"
    >
      const
    </span>
     test = 
    <span
      className={
        Array [
          "hljs-number",
        ]
      }
      key="lo-0-2"
    >
      1
    </span>
    ;
  </code>
</pre>
`;

exports[`Code renders multiline correctly 1`] = `
<pre
  className=""
>
  <code>
    <span
      className={
        Array [
          "hljs-keyword",
        ]
      }
      key="lo-0-0"
    >
      const
    </span>
     test = 
    <span
      className={
        Array [
          "hljs-number",
        ]
      }
      key="lo-0-2"
    >
      1
    </span>
    ;

    <span
      className={
        Array [
          "hljs-keyword",
        ]
      }
      key="lo-0-4"
    >
      const
    </span>
     f = 
    <span
      className={
        Array [
          "hljs-function",
        ]
      }
      key="lo-0-6"
    >
      (
      <span
        className={
          Array [
            "hljs-params",
          ]
        }
        key="lo-1-1"
      >
        t
      </span>
      ) =&gt;
    </span>
     t + 
    <span
      className={
        Array [
          "hljs-number",
        ]
      }
      key="lo-0-8"
    >
      1
    </span>
    ;
  </code>
</pre>
`;
